src/systemctl/systemctl-set-environment.c

```diff
@@ -42,14 +80,25 @@ int show_environment(int argc, char *argv[], void *userdata) {
         if (r < 0)
                 return log_error_errno(r, "Failed to get environment: %s", bus_error_message(&error, r));
 
-        r = sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "s");
-        if (r < 0)
-                return bus_log_parse_error(r);
+        if (OUTPUT_MODE_IS_JSON(arg_output)) {
+                _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+                JsonFormatFlags flags = output_mode_to_json_format_flags(arg_output);
 
-        while ((r = sd_bus_message_read_basic(reply, SD_BUS_TYPE_STRING, &text)) > 0) {
-                r = print_variable(text);
+                r = json_transform_message(reply, &v);
                 if (r < 0)
                         return r;
+
+                json_variant_dump(v, flags, stdout, NULL);
+        } else {
+                r = sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "s");
```

Comment:
this is common for both paths, so just do it outside and remove it from the function